# Tarea 2: Gráficos estadísticos interactivos con plotly
## Tema: Desastres en Nicaragua (1992-2013)

### Introducción
La base de datos _DesInventar_ recopila noticias sobre desastres por amenazas naturales o antropogénicas en América Latina. Específicamente, el "desarrollo de DesInventar, con una concepción que permite ver a los desastres desde una escala espacial local (municipio o equivalente),facilita diálogos para gestión de riesgos entre actores e instituciones y sectores, y con gobiernos provinciales y nacionales" ([DesInventar](https://www.desinventar.org/), 2020). Para este ejercicio, se utilizará la base de datos de ocurrencia de desastres para Nicaragua entre 1992 y 2013. 

**Objetivos:**
1. Crear un gráfico interactivo que muestre la tendencia temporal de los desastres en Nicaragua.
2. Realizar un gráfico interactivo de los 10 municipios con más desastres a lo largo del tiempo. 
3. Graficar interactivamente la distribución de los 10 desastres más comunes.

-----------

## Carga de paquetes

In [1]:
# Carga de pandas con el alias pd
import pandas as pd

# Carga de plotly.express con el alias px
import plotly.express as px

# Carga de plotly.graph_objects con el alias go
import plotly.graph_objects as go

# Configuración de pandas para mostrar separadores de miles y 2 dígitos decimales
pd.set_option('display.float_format', '{:,.0f}'.format)

<p>

## Carga de datos desde el repositorio en GitHub

In [2]:
# Carga de datos en el dataframe 'desastres' con la base de datos de DesInventar
desastres = pd.read_csv("https://raw.githubusercontent.com/gabs18/Tarea2_GF0657/refs/heads/main/DESINVENTAR_NICARAGUA.csv", sep=";")

# Emplear la columna 'Serial' como el índice único
desastres.set_index('Serial', inplace=True)

# Despliegue de una muestra aleatoria de 5 filas
desastres.sample(5)

Año    Evento  ADM1_CODE           ADM1  ADM2_CODE        ADM2  \
Serial                                                                        
2013-00146 2,013     FLOOD         16           RAAS      1,608  Bluefields   
2011-00053 2,011     FLOOD          1  Nueva Segovia        106     Dipilto   
2012-00098 2,012     SPATE         14      Matagalpa      1,402  Río Blanco   
2000-00055 2,000      RAIN          5           León        502    El Sauce   
1996-00004 1,996  EPIDEMIC         14      Matagalpa      1,408    Matiguas   

                                                        LUGAR       Fecha  \
Serial                                                                      
2013-00146  Comunidad San Sebastián en Caño Negro. Barrios...   2013/7/22   
2011-00053                                          Río Murra   2011/10/4   
2012-00098  Carretera de Río Blanco hacia Mulukukú, sector...  2012/10/21   
2000-00055                         En la comunidad El Jocotal   2000/10/3   
1996-00004                             Municipio de Matiguás.   1996/1/12   

                                                Observaciones          Causa  \
Serial                                                                         
2013-00146  Tres casas colapsaron e el centro de la ciudad...           RAIN   
2011-00053  Dos personas murieron, arrastradas por las cor...       OVERFLOW   
2012-00098  Buses y camiones de transporte y de carga en h...           RAIN   
2000-00055  Los datos de damnificados se dedujo de multipl...        Huracán   
1996-00004  Pobladores del Municipio de Matiguás, toman ag...  CONTAMINATION   

            ... Viviendas Afectadas Damnificados Afectados  Reubicados  \
Serial      ...                                                          
2013-00146  ...                   7           15        95         NaN   
2011-00053  ...                 NaN          NaN       NaN         NaN   
2012-00098  ...                 NaN          NaN       NaN         NaN   
2000-00055  ...                 NaN            5       NaN         NaN   
1996-00004  ...                 NaN          NaN       NaN         NaN   

           Evacuados  Pérdidas $USD  Pérdidas $Local  Centros Educativos  \
Serial                                                                     
2013-00146       NaN            NaN              NaN                 NaN   
2011-00053       NaN            NaN              NaN                 NaN   
2012-00098       NaN            NaN              NaN                 NaN   
2000-00055       NaN            NaN              NaN                 NaN   
1996-00004       NaN            NaN              NaN                 NaN   

            Centros Médicos  Daños cultivos Ha  
Serial                                          
2013-00146              NaN                NaN  
2011-00053              NaN                NaN  
2012-00098              NaN                NaN  
2000-00055              NaN                NaN  
1996-00004              NaN                NaN  

[5 rows x 30 columns]

<p>

## Gráficos

### Línea temporal de desastres en Nicaragua: **gráfico de línea**

In [3]:
# Contar los desastres por año
def contar_desastres_por_anio(desastres):
    # Agrupar por 'año' y contar el número de desastres en cada año
    conteo_desastres = desastres.groupby('Año').size().reset_index(name='total_desastres')
    # Ordenar por año para mejorar la visualización
    conteo_desastres = conteo_desastres.sort_values(by='Año')
    return conteo_desastres

# Ejecutar la función
conteo_desastres_anio = contar_desastres_por_anio(desastres)
print(conteo_desastres_anio)

     Año  total_desastres
0  1,992               33
1  1,993               34
2  1,994               23
3  1,995               46
4  1,996               68
5  1,997               31
6  1,998               49
7  1,999               55
8  2,000               61
9  2,001               70
10 2,002               29
11 2,003               46
12 2,004               29
13 2,005               50
14 2,006               17
15 2,007               95
16 2,008              115
17 2,009               54
18 2,010               81
19 2,011               65
20 2,012              115
21 2,013              232


In [4]:
# Crear gráfico de líneas para mostrar la evolución temporal
fig = px.line(conteo_desastres_anio, x='Año', y='total_desastres',
              title='Evolución temporal de desastres por año en Nicaragua',
              template='ygridoff',
              labels={'año': 'Año', 'total_desastres': 'Cantidad de desastres'}, 
              markers=True)

# Atributos de visualización de la figura
fig.update_layout(
    title=dict(
        x=0.5,  # Centrar el título
        font=dict(size=20)),
    xaxis=dict(showgrid=True, gridwidth=0.5, gridcolor='lightgray'),
    yaxis=dict(showgrid=True, gridwidth=0.5, gridcolor='lightgray')
)

fig.show()

El gráfico anterior muestra la evolución de cantidad de desastres registrados entre 1992 y 2013 para Nicaraguaa nivel nacional. Se logra evidenciar un patrón de aumento de desastres a partir del año 2007. De hecho, el 30 de mayo del 2008 la tormenta tropical Alma impactó el territorio nicaragüense cerca de la ciudad de León y afectó los departamentos: León Chinandega, Managua, Masaya, Carazo, Granada y Rivas (OCHA, 2008). El 2013 se registraron hasta 232 desastres, y que el que dio más afectación fue una contaminación de una planta de agua potable: 50.000 personas fueron afectadas según la base de datos DesInventar.

<p>

### Línea temporal por año y por municipio: **gráfico de líneas**

In [20]:
# Función para contar desastres por municipio y año
def contar_desastres_por_municipio_anio(desastres):
    # Agrupar por 'municipio' y 'año', y contar el número de desastres
    conteo_desastres = desastres.groupby(['ADM2', 'Año']).size().reset_index(name='total_desastres_adm2')
    # Ordenar por municipio y año
    conteo_desastres = conteo_desastres.sort_values(by=['ADM2', 'Año'])
    return conteo_desastres

# Obtener el conteo de desastres por municipio y año
conteo_desastres_municipio_anio = contar_desastres_por_municipio_anio(desastres)
print(conteo_desastres_municipio_anio)

           ADM2   Año  total_desastres_adm2
0       Achuapa 2,000                     1
1       Achuapa 2,012                     1
2       Acoyapa 2,001                     1
3       Acoyapa 2,013                     2
4    Altagracia 1,996                     2
..          ...   ...                   ...
788      Wiwilí 2,008                     2
789      Wiwilí 2,013                     1
790   Yalagüina 1,992                     1
791   Yalagüina 2,012                     1
792   Yalagüina 2,013                     1

[793 rows x 3 columns]


In [ ]:
# Obtener el total de desastres por municipio
total_desastres_por_municipio = conteo_desastres_municipio_anio.groupby('ADM2')['total_desastres_adm2'].sum().reset_index()

# Seleccionar los 10 municipios con mayor cantidad de desastres
top_10_municipios = total_desastres_por_municipio.nlargest(10, 'total_desastres_adm2')['ADM2']

# Filtrar los datos para incluir solo los municipios del top 15
conteo_top_10 = conteo_desastres_municipio_anio[conteo_desastres_municipio_anio['ADM2'].isin(top_10_municipios)]

# Crear el gráfico de líneas con Plotly
fig = px.line(conteo_top_10, x='Año', y='total_desastres_adm2', color='ADM2',
              title='Evolución temporal de desastres en los 10 municipios más afectados de Nicaragua',
              labels={'Año': 'Año', 'total_desastres_adm2': 'Cantidad de desastres', 'ADM2': 'Municipio'},
              template='plotly_white',
              markers=True)

# Atributos de visualización de la figura
fig.update_layout(
    title=dict(
        x=0.5,  # Centrar el título
        font=dict(size=20)),
    xaxis=dict(showgrid=True, gridwidth=0.5, gridcolor='lightgray'),
    yaxis=dict(showgrid=True, gridwidth=0.5, gridcolor='lightgray')
)

fig.show()

El gráfico de líneas anterior tiene la función de mostrar los patrones de cantidad de desastres por municipio. Es complementario al primero, pero con la ventaja de discernir cuáles partes de Nicaragua han sido más afectadas. Resalta Managua, León y Bilwi, municipios con alta población y cercanía a cuerpos de agua. Según la base de datos, Bilwi en la costa caribeña, ha sido afectado mayoritariamente por inundaciones y tormentas.

<p>

### Los desastres más comunes: **gráfico de barras**

In [ ]:
# Contar los desastres por tipo de evento
def obtener_top_eventos(df, top_n=10):
    conteo_evento = df['Evento'].value_counts().reset_index()
    conteo_evento.columns = ['Evento', 'Total_Desastres']
    # Seleccionar los top_n eventos con mayor cantidad de desastres
    top_eventos = conteo_evento.nlargest(top_n, 'Total_Desastres')
    return top_eventos

# Obtener los 10 tipos de desastres con mayor cantidad
top_10_eventos = obtener_top_eventos(desastres)


# Traducción de términos con un diccionario de traducciones
traducciones = {'LANDSLIDE': 'Deslizamiento', 
                'STRONGWIND': 'Vientos fuertes', 
                'RAIN': 'Lluvia', 
                'POLLUTION': 'Contaminación',
                'SPATE': 'Avalancha',
                'FORESTFIRE': 'Incendio forestal',
                'EPIDEMIC': 'Epidemia',
                'DROUGHT': 'Sequía',
                'FIRE': 'Incendio',
                'FLOOD': 'Inundación'
}

# Reemplazar celdas en la columna 'Evento' con los valores del diccionario
top_10_eventos['Evento'] = top_10_eventos['Evento'].replace(traducciones)

top_10_eventos.sample(5)


Evento  Total_Desastres
2         Sequía              112
5      Avalancha               83
7         Lluvia               54
6  Contaminación               58
0     Inundación              467

In [9]:
# Crear el gráfico de barras con Plotly
fig = px.bar(top_10_eventos, x='Evento', y='Total_Desastres',
             title='Top 10 de desastres por tipo de evento (Nicaragua: 1992-2013)',
             labels={'Evento': 'Tipo de desastre', 'Total_Desastres': 'Cantidad de desastres'},
             template='plotly_white',
             color='Evento', color_discrete_sequence=px.colors.sequential.haline)

# Atributos de visualización de la figura
fig.update_layout(
    title=dict(
        x=0.5,  # Centrar el título
        font=dict(size=20)),
    xaxis={'categoryorder':'total ascending'},
    yaxis=dict(showgrid=True, gridwidth=0.5, gridcolor='lightgray')
)

fig.show()

El último gráfico tiene el objetivo de evidenciar cuáles han sido los desastres más recurrentes: inundaciones, incendios y sequías. Esos tres dsastres suman un total de 695 desastres entre 1992 y 2013. Según el INETER (2018), Nicaragua se puede catalogar como uno de los países con mayor riesgo a desastres por su "conformación geológica, estructura morfológica, ubicación en el istmo Centroamericano, cuencas hidrográficas, régimen de las lluvias, [...] la distribución y desarrollo económico, social y cultural de la población". El mismo instito indica que hay una alta susceptibilidad a inundaciones, huracanes, deslizamientos, sismos, erupciones volcánicas, sequías, tsunamis e incendios forestales; y que esto se junta con el grado de exposición de la población y las vulnerabilidades socioeconómicas. Por lo tanto, el gráfico de barras generado con datos de la base DesInventar ayuda a cuantificar cuáles han sido los eventos que más han ocurrido y potencialmente afectado a la población.

------

### Referencias bibliográficas
- INETER. (2018). Portal de Mapas Interactivos de Vulnerabilidades y Amenazas Naturales. _Ineter.gob.ni._ https://www.ineter.gob.ni/geoportales/vulnerabilidades-amenazas/index.html

- OCHA. (2008, 30 de mayo). Nicaragua: Tormenta ALMA afecta la zona del occidente. _ReliefWeb._ https://reliefweb.int/report/nicaragua/nicaragua-tormenta-alma-afecta-la-zona-del-occidente


‌

‌